sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,380 @@
+
+  @transient private val lock = new Object()
+
+  // The logical plan optimizer for re-optimizing the current logical plan.
+  private object Optimizer extends RuleExecutor[LogicalPlan] {
+    // TODO add more optimization rules
+    override protected def batches: Seq[Batch] = Seq()
+  }
+
+  // A list of physical plan rules to be applied before creation of query stages. The physical
+  // plan should reach a final status of query stages (i.e., no more addition or removal of
+  // Exchange nodes) after running these rules.
+  @transient private val queryStagePreparationRules: Seq[Rule[SparkPlan]] = Seq(
+    PlanAdaptiveSubqueries(subqueryMap),
+    EnsureRequirements(conf)
+  )
+
+  // A list of physical optimizer rules to be applied to a new stage before its execution. These
+  // optimizations should be stage-independent.
+  @transient private val queryStageOptimizerRules: Seq[Rule[SparkPlan]] = Seq(
+    CollapseCodegenStages(conf)
+  )
+
+  private var currentStageId = 0
+
+  @volatile private var currentPhysicalPlan =
+    applyPhysicalRules(initialPlan, queryStagePreparationRules)
+
+  /**
+   * Return type for `createQueryStages`
+   * @param newPlan the new plan with created query stages.
+   * @param allChildStagesMaterialized whether all child stages have been materialized.
+   * @param newStages the newly created query stages, including new reused query stages.
+   */
+  private case class CreateStageResult(
+    newPlan: SparkPlan,
+    allChildStagesMaterialized: Boolean,
+    newStages: Seq[(Exchange, QueryStageExec)])
+
+  /**
+   * This method is called recursively to traverse the plan tree bottom-up and checks whether or
+   * not the child query stages (if any) of the current node have all been materialized. If the
+   * current node is:
+   * 1. An Exchange:
+   *    1) If the Exchange is equivalent to another Exchange of an existing query stage, create
+   *       a reused query stage and replace the current node with the reused stage.
+   *    2) If all its child query stages have been materialized, create a new query stage and
+   *       replace the current node with the new stage.
+   *    3) Otherwise, do nothing.
+   *    In any of the above cases, the method will return `false` for `allChildStagesMaterialized`.
+   * 2. A query stage:
+   *    return `true` for `allChildStagesMaterialized` if this stage is completed, otherwise
+   *    `false`.
+   * 3. Recursively visit all child nodes, and propagate the return value.
+   */
+  private def createQueryStages(plan: SparkPlan): CreateStageResult = plan match {
+    case e: Exchange =>
+      stageCache.get(e.canonicalized) match {
+        case Some(existingStage) if conf.exchangeReuseEnabled =>
+          val reusedStage = reuseQueryStage(existingStage, e.output)
+          CreateStageResult(newPlan = reusedStage,
+            allChildStagesMaterialized = false, newStages = Seq((e, reusedStage)))
```

Comment:
if the reused Exchange is already materialized, shall we return true here?

Reply:
We could, but we might lose the opportunity of optimization that depends on the stats of a reused stage. And since this exchange reuse is entirely dynamic, choosing to optimize for new stages but not for completed reused stages will make the adaptive plan even less deterministic.